python 3 master <==> python 2 worker, does not work

From the buildbot master branch, I experimented with setting up
a Python 3.6 master and a Python 2.7 worker.  I was able to start the master and worker,
and saw that they connected to each other.

However, when I triggered a build, I got:

```
Traceback (most recent call last):
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 653, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1442, in gotResult
    _inlineCallbacks(r, g, deferred)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
--- <exception caught here> ---
  File "/home/vagrant/buildbot/master/buildbot/process/buildstep.py", line 574, in startStep
    self.results = yield self.run()
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
    result = g.send(result)
  File "/home/vagrant/buildbot/master/buildbot/process/buildstep.py", line 704, in run
    results = yield self._start_deferred
  File "/home/vagrant/buildbot/master/buildbot/steps/source/git.py", line 171, in startVC
    gitInstalled = yield self.checkBranchSupport()
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/vagrant/buildbot/master/buildbot/steps/source/git.py", line 583, in checkBranchSupport
    stdout = yield self._dovccmd(['--version'], collectStdout=True)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/vagrant/buildbot/master/buildbot/steps/source/git.py", line 381, in _dovccmd
    yield self.runCommand(cmd)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/vagrant/buildbot-venv/lib64/python3.6/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/vagrant/buildbot/master/buildbot/process/buildstep.py", line 900, in runCommand
    res = yield command.run(self, self.remote, self.build.builder.name)
buildbot.process.remotecommand.RemoteException: b'exceptions.AssertionError': b"Unexpected usePTY argument value: u'slave-config'. Expected boolean."
b'Traceback (most recent call last):\n  File "/home/vagrant/buildbot-venv-2.7/lib/python2.7/site-packages/twisted/spread/pb.py", line 1054, in _recvMessage\n    netResult = object.remoteMessageReceived(self, message, netArgs, netKw)\n  File "/home/vagrant/buildbot-venv-2.7/lib/python2.7/site-packages/twisted/spread/flavors.py", line 125, in remoteMessageReceived\n    state = method(*args, **kw)\n  File "/home/vagrant/buildbot/worker/buildbot_worker/base.py", line 150, in remote_startCommand\n    d = self.command.doStart()\n  File "/home/vagrant/buildbot/worker/buildbot_worker/commands/base.py", line 163, in doStart\n    d = defer.maybeDeferred(self.start)\n--- <exception caught here> ---\n  File "/home/vagrant/buildbot-venv-2.7/lib/python2.7/site-packages/twisted/internet/defer.py", line 150, in maybeDeferred\n    result = f(*args, **kw)\n  File "/home/vagrant/buildbot/worker/buildbot_worker/commands/shell.py", line 47, in start\n    logEnviron=args.get(\'logEnviron\', True),\n  File "/home/vagrant/buildbot/worker/buildbot_worker/runprocess.py", line 388, in __init__\n    usePTY)\nexceptions.AssertionError: Unexpected usePTY argument value: u\'slave-config\'. Expected boolean.\n'
```

--- a/master/buildbot/pbutil.py
+++ b/master/buildbot/pbutil.py
@@ -19,11 +19,14 @@
 
 from __future__ import absolute_import
 from __future__ import print_function
+from future.utils import iteritems
 
 from twisted.internet import protocol
 from twisted.python import log
 from twisted.spread import pb
 from twisted.spread.pb import PBClientFactory
+
+from buildbot.util import bytes2unicode
 
 
 class NewCredPerspective(pb.Avatar):
@@ -156,3 +159,19 @@ class ReconnectingPBClientFactory(PBClientFactory,
         # probably authorization
         self.stopTrying()  # logging in harder won't help
         log.err(why)
+
+
+def decode(data, encoding='utf-8', errors='strict'):
+    """We need to convert a dictionary where keys and values
+    are bytes, to unicode strings.  This happens when a
+    Python 2 worker sends a dictionary back to a Python 3 master.
+    """
+    data_type = type(data)
+
+    if data_type == bytes:
+        return bytes2unicode(data, encoding, errors)
+    if data_type in (dict, list, tuple):
+        if data_type == dict:
+            data = iteritems(data)
+        return data_type(map(decode, data))
+    return data

--- a/master/buildbot/process/remotecommand.py
+++ b/master/buildbot/process/remotecommand.py
@@ -25,6 +25,7 @@ from twisted.python.failure import Failure
 from twisted.spread import pb
 
 from buildbot import util
+from buildbot.pbutil import decode
 from buildbot.process import metrics
 from buildbot.process.results import FAILURE
 from buildbot.process.results import SUCCESS
@@ -189,6 +190,7 @@ class RemoteCommand(base.RemoteCommandImpl, WorkerAPICompatMixin):
         @type  updates: list of [object, int]
         @param updates: list of updates from the remote command
         """
+        updates = decode(updates)
         self.worker.messageReceivedFromWorker()
         max_updatenum = 0
         for (update, num) in updates:

--- a/master/buildbot/test/integration/test_usePty.py
+++ b/master/buildbot/test/integration/test_usePty.py
@@ -16,6 +16,9 @@
 from __future__ import absolute_import
 from __future__ import print_function
 
+from distutils.version import LooseVersion
+
+from twisted import __version__ as twistedVersion
 from twisted.internet import defer
 
 from buildbot.test.util.decorators import skipUnlessPlatformIs
@@ -35,6 +38,16 @@ class ShellMaster(RunMasterBase):
         self.assertEqual(build['buildid'], 1)
         res = yield self.checkBuildStepLogExist(build, "in a terminal", onlyStdout=True)
         self.assertTrue(res)
+
+        # Twisted versions less than 17.1.0 would issue a warning:
+        #    "Argument strings and environment keys/values passed to reactor.spawnProcess
+        #    "should be str, not unicode."
+        # This warning was unnecessary. Even in the old versions of Twisted, the
+        # unicode arguments were encoded.  This warning was removed in Twisted here:
+        #
+        # https://github.com/twisted/twisted/commit/23fa3cc05549251ea4118e4e03354d58df87eaaa
+        if LooseVersion(twistedVersion) < LooseVersion("17.1.0"):
+            self.flushWarnings()
 
     @skipUnlessPlatformIs('posix')
     @defer.inlineCallbacks

--- a/master/buildbot/worker/protocols/pb.py
+++ b/master/buildbot/worker/protocols/pb.py
@@ -25,6 +25,7 @@ from twisted.internet import reactor
 from twisted.python import log
 from twisted.spread import pb
 
+from buildbot.pbutil import decode
 from buildbot.worker.protocols import base
 
 
@@ -199,7 +200,7 @@ class Connection(base.Connection, pb.Avatar):
             with _wrapRemoteException():
                 # Try to call buildbot-worker method.
                 info = yield self.mind.callRemote('getWorkerInfo')
-            defer.returnValue(info)
+            defer.returnValue(decode(info))
         except _NoSuchMethod:
             yield self.remotePrint(
                 "buildbot-slave detected, failing back to deprecated buildslave API. "
@@ -237,7 +238,7 @@ class Connection(base.Connection, pb.Avatar):
             except _NoSuchMethod:
                 log.msg("Worker.getVersion is unavailable - ignoring")
 
-            defer.returnValue(info)
+            defer.returnValue(decode(info))
 
     def remoteSetBuilderList(self, builders):
         d = self.mind.callRemote('setBuilderList', builders)

--- a/worker/buildbot_worker/base.py
+++ b/worker/buildbot_worker/base.py
@@ -31,7 +31,7 @@ import buildbot_worker
 from buildbot_worker import monkeypatches
 from buildbot_worker.commands import base
 from buildbot_worker.commands import registry
-from buildbot_worker.compat import bytes2NativeString
+from buildbot_worker.compat import bytes2unicode
 
 
 class UnknownCommand(pb.Error):
@@ -80,8 +80,8 @@ class WorkerForBuilderBase(service.Service):
     def setBuilddir(self, builddir):
         assert self.parent
         self.builddir = builddir
-        self.basedir = os.path.join(bytes2NativeString(self.bot.basedir),
-                                    bytes2NativeString(self.builddir))
+        self.basedir = os.path.join(bytes2unicode(self.bot.basedir),
+                                    bytes2unicode(self.builddir))
         if not os.path.isdir(self.basedir):
             os.makedirs(self.basedir)
 

--- a/worker/buildbot_worker/base.py
+++ b/worker/buildbot_worker/base.py
@@ -32,6 +32,7 @@ from buildbot_worker import monkeypatches
 from buildbot_worker.commands import base
 from buildbot_worker.commands import registry
 from buildbot_worker.compat import bytes2unicode
+from buildbot_worker.pbutil import decode
 
 
 class UnknownCommand(pb.Error):
@@ -131,6 +132,9 @@ class WorkerForBuilderBase(service.Service):
         .commandComplete() to notify the master-side RemoteCommand that I'm
         done.
         """
+        stepId = decode(stepId)
+        command = decode(command)
+        args = decode(args)
 
         self.activity()
 
@@ -141,10 +145,10 @@ class WorkerForBuilderBase(service.Service):
         try:
             factory = registry.getFactory(command)
         except KeyError:
-            raise UnknownCommand("unrecognized WorkerCommand '{0}'".format(command))
+            raise UnknownCommand(u"unrecognized WorkerCommand '{0}'".format(command))
         self.command = factory(self, stepId, args)
 
-        log.msg(" startCommand:{0} [id {1}]".format(command, stepId))
+        log.msg(u" startCommand:{0} [id {1}]".format(command, stepId))
         self.remoteStep = stepref
         self.remoteStep.notifyOnDisconnect(self.lostRemoteStep)
         d = self.command.doStart()

--- a/worker/buildbot_worker/base.py
+++ b/worker/buildbot_worker/base.py
@@ -67,7 +67,7 @@ class WorkerForBuilderBase(service.Service):
         self.setName(name)
 
     def __repr__(self):
-        return "<WorkerForBuilder '%s' at %d>" % (self.name, id(self))
+        return "<WorkerForBuilder '{0}' at {1}>".format(self.name, id(self))
 
     def setServiceParent(self, parent):
         service.Service.setServiceParent(self, parent)
@@ -103,8 +103,8 @@ class WorkerForBuilderBase(service.Service):
         self.remote.notifyOnDisconnect(self.lostRemote)
 
     def remote_print(self, message):
-        log.msg("WorkerForBuilder.remote_print(%s): message from master: %s" %
-                (self.name, message))
+        log.msg("WorkerForBuilder.remote_print({0}): message from master: {1}".format(
+                self.name, message))
 
     def lostRemote(self, remote):
         log.msg("lost remote")
@@ -141,10 +141,10 @@ class WorkerForBuilderBase(service.Service):
         try:
             factory = registry.getFactory(command)
         except KeyError:
-            raise UnknownCommand("unrecognized WorkerCommand '%s'" % command)
+            raise UnknownCommand("unrecognized WorkerCommand '{0}'".format(command))
         self.command = factory(self, stepId, args)
 
-        log.msg(" startCommand:%s [id %s]" % (command, stepId))
+        log.msg(" startCommand:{0} [id {1}]".format(command, stepId))
         self.remoteStep = stepref
         self.remoteStep.notifyOnDisconnect(self.lostRemoteStep)
         d = self.command.doStart()
@@ -154,7 +154,7 @@ class WorkerForBuilderBase(service.Service):
 
     def remote_interruptCommand(self, stepId, why):
         """Halt the current step."""
-        log.msg("asked to interrupt current command: %s" % why)
+        log.msg("asked to interrupt current command: {0}".format(why))
         self.activity()
         if not self.command:
             # TODO: just log it, a race could result in their interrupting a
@@ -170,7 +170,7 @@ class WorkerForBuilderBase(service.Service):
         silence it, and then forget about it."""
         if not self.command:
             return
-        log.msg("stopCommand: halting current command %s" % self.command)
+        log.msg("stopCommand: halting current command {0}".format(self.command))
         self.command.doInterrupt()  # shut up! and die!
         self.command = None  # forget you!
 
@@ -266,8 +266,8 @@ class BotBase(service.MultiService):
             b = self.builders.get(name, None)
             if b:
                 if b.builddir != builddir:
-                    log.msg("changing builddir for builder %s from %s to %s"
-                            % (name, b.builddir, builddir))
+                    log.msg("changing builddir for builder {0} from {1} to {2}".format(
+                            name, b.builddir, builddir))
                     b.setBuilddir(builddir)
             else:
                 b = self.WorkerForBuilder(name)
@@ -292,9 +292,9 @@ class BotBase(service.MultiService):
         for dir in os.listdir(self.basedir):
             if os.path.isdir(os.path.join(self.basedir, dir)):
                 if dir not in wanted_dirs:
-                    log.msg("I have a leftover directory '%s' that is not "
+                    log.msg("I have a leftover directory '{0}' that is not "
                             "being used by the buildmaster: you can delete "
-                            "it now" % dir)
+                            "it now".format(dir))
 
         defer.returnValue(retval)
 
@@ -365,7 +365,7 @@ class WorkerBase(service.MultiService):
         # first, apply all monkeypatches
         monkeypatches.patch_all()
 
-        log.msg("Starting Worker -- version: %s" % buildbot_worker.version)
+        log.msg("Starting Worker -- version: {0}".format(buildbot_worker.version))
 
         if self.umask is not None:
             os.umask(self.umask)
@@ -388,6 +388,6 @@ class WorkerBase(service.MultiService):
 
         try:
             with open(filename, "w") as f:
-                f.write("%s\n" % hostname)
+                f.write("{0}\n".format(hostname))
         except Exception:
             log.msg("failed - ignoring")

--- a/worker/buildbot_worker/compat.py
+++ b/worker/buildbot_worker/compat.py
@@ -56,22 +56,40 @@ def bytes2NativeString(x, encoding='utf-8'):
     return x
 
 
-def unicode2bytes(x, encoding='utf-8'):
+def unicode2bytes(x, encoding='utf-8', errors='strict'):
     """
     Convert a unicode string to C{bytes}.
 
     @param x: a unicode string, of type C{unicode} on Python 2,
               or C{str} on Python 3.
     @param encoding: an optional codec, default: 'utf-8'
+    @param errors: error handling scheme, default 'strict'
     @return: a string of type C{bytes}
     """
     if isinstance(x, text_type):
-        x = x.encode(encoding)
+        x = x.encode(encoding, errors)
     return x
+
+
+def bytes2unicode(x, encoding='utf-8', errors='strict'):
+    """
+    Convert a C{bytes} to a unicode string.
+
+    @param x: a unicode string, of type C{unicode} on Python 2,
+              or C{str} on Python 3.
+    @param encoding: an optional codec, default: 'utf-8'
+    @param errors: error handling scheme, default 'strict'
+    @return: a unicode string of type C{unicode} on Python 2, or
+             C{str} on Python 3.
+    """
+    if isinstance(x, (text_type, type(None))):
+        return x
+    return text_type(x, encoding, errors)
 
 
 __all__ = [
     "NativeStringIO",
     "bytes2NativeString",
+    "bytes2unicode",
     "unicode2bytes"
 ]

--- a/worker/buildbot_worker/pbutil.py
+++ b/worker/buildbot_worker/pbutil.py
@@ -19,6 +19,7 @@
 
 from __future__ import absolute_import
 from __future__ import print_function
+from future.utils import iteritems
 
 from twisted.cred import error
 from twisted.internet import protocol
@@ -26,6 +27,8 @@ from twisted.internet import reactor
 from twisted.python import log
 from twisted.spread import pb
 from twisted.spread.pb import PBClientFactory
+
+from buildbot_worker.compat import bytes2unicode
 
 
 class ReconnectingPBClientFactory(PBClientFactory,
@@ -123,3 +126,19 @@ class ReconnectingPBClientFactory(PBClientFactory,
 
         # lose the current connection, which will trigger a retry
         broker.transport.loseConnection()
+
+
+def decode(data, encoding='utf-8', errors='strict'):
+    """We need to convert a dictionary where keys and values
+    are bytes, to unicode strings.  This happens when a
+    Python 2 master sends a dictionary back to a Python 3 worker.
+    """
+    data_type = type(data)
+
+    if data_type == bytes:
+        return bytes2unicode(data, encoding, errors)
+    if data_type in (dict, list, tuple):
+        if data_type == dict:
+            data = iteritems(data)
+        return data_type(map(decode, data))
+    return data

--- a/worker/buildbot_worker/runprocess.py
+++ b/worker/buildbot_worker/runprocess.py
@@ -101,7 +101,7 @@ class LogFileWatcher(object):
         self.name = name
         self.logfile = logfile
 
-        log.msg("LogFileWatcher created to watch %s" % logfile)
+        log.msg("LogFileWatcher created to watch {0}".format(logfile))
         # we are created before the ShellCommand starts. If the logfile we're
         # supposed to be watching already exists, record its size and
         # ctime/mtime so we can tell when it starts to change.
@@ -195,8 +195,8 @@ class RunProcessPP(protocol.ProcessProtocol):
 
         if self.command.useProcGroup:
             if self.debug:
-                log.msg(" recording pid %d as subprocess pgid"
-                        % (self.transport.pid,))
+                log.msg(" recording pid {0} as subprocess pgid".format(
+                    self.transport.pid))
             self.transport.pgid = self.transport.pid
 
         if self.pending_stdin:
@@ -364,7 +364,7 @@ class RunProcess(object):
                 if v is not None:
                     if not isinstance(v, string_types):
                         raise RuntimeError("'env' values must be strings or "
-                                           "lists; key '%s' is incorrect" % (key,))
+                                           "lists; key '{0}' is incorrect".format(key))
                     newenv[key] = p.sub(subst, v)
 
             self.environ = newenv
@@ -425,7 +425,7 @@ class RunProcess(object):
             self.logFileWatchers.append(w)
 
     def __repr__(self):
-        return "<%s '%s'>" % (self.__class__.__name__, self.fake_command)
+        return "<{0} '{1}'>".format(self.__class__.__name__, self.fake_command)
 
     def sendStatus(self, status):
         self.builder.sendUpdate(status)
@@ -455,8 +455,8 @@ class RunProcess(object):
             os.makedirs(self.workdir)
         log.msg("RunProcess._startCommand")
         if self.notreally:
-            self._addToBuffers('header', "command '%s' in dir %s" %
-                               (self.fake_command, self.workdir))
+            self._addToBuffers('header', "command '{0}' in dir {1}".format(
+                               self.fake_command, self.workdir))
             self._addToBuffers('header', "(not really)\n")
             self.finished(None, 0)
             return
@@ -512,28 +512,28 @@ class RunProcess(object):
         self._addToBuffers('header', display + "\n")
 
         # then comes the secondary information
-        msg = " in dir %s" % (self.workdir,)
+        msg = " in dir {0}".format(self.workdir)
         if self.timeout:
             if self.timeout == 1:
                 unit = "sec"
             else:
                 unit = "secs"
-            msg += " (timeout %d %s)" % (self.timeout, unit)
+            msg += " (timeout {0} {1})".format(self.timeout, unit)
         if self.maxTime:
             if self.maxTime == 1:
                 unit = "sec"
             else:
                 unit = "secs"
-            msg += " (maxTime %d %s)" % (self.maxTime, unit)
+            msg += " (maxTime {0} {1})".format(self.maxTime, unit)
         log.msg(" " + msg)
         self._addToBuffers('header', msg + "\n")
 
-        msg = " watching logfiles %s" % (self.logfiles,)
+        msg = " watching logfiles {0}".format(self.logfiles)
         log.msg(" " + msg)
         self._addToBuffers('header', msg + "\n")
 
         # then the obfuscated command array for resolving unambiguity
-        msg = " argv: %s" % (self.fake_command,)
+        msg = " argv: {0}".format(self.fake_command)
         log.msg(" " + msg)
         self._addToBuffers('header', msg + "\n")
 
@@ -542,16 +542,16 @@ class RunProcess(object):
             msg = " environment:\n"
             env_names = sorted(self.environ.keys())
             for name in env_names:
-                msg += "  %s=%s\n" % (name, self.environ[name])
-            log.msg(" environment:\n%s" % (pprint.pformat(self.environ),))
+                msg += "  {0}={1}\n".format(name, self.environ[name])
+            log.msg(" environment:\n{0}".format(pprint.pformat(self.environ)))
             self._addToBuffers('header', msg)
 
         if self.initialStdin:
-            msg = " writing %d bytes to stdin" % len(self.initialStdin)
+            msg = " writing {0} bytes to stdin".format(len(self.initialStdin))
             log.msg(" " + msg)
             self._addToBuffers('header', msg + "\n")
 
-        msg = " using PTY: %s" % bool(self.usePTY)
+        msg = " using PTY: {0}".format(bool(self.usePTY))
         log.msg(" " + msg)
         self._addToBuffers('header', msg + "\n")
 
@@ -766,7 +766,7 @@ class RunProcess(object):
 
     def finished(self, sig, rc):
         self.elapsedTime = util.now(self._reactor) - self.startTime
-        log.msg("command finished with signal %s, exit code %s, elapsedTime: %0.6f" % (
+        log.msg("command finished with signal {0}, exit code {1}, elapsedTime: {2:0.6f}".format(
             sig, rc, self.elapsedTime))
         for w in self.logFileWatchers:
             # this will send the final updates
@@ -777,7 +777,7 @@ class RunProcess(object):
         if self.sendRC:
             if sig is not None:
                 self.sendStatus(
-                    {'header': "process killed by signal %d\n" % sig})
+                    {'header': "process killed by signal {0}\n".format(sig)})
             self.sendStatus({'rc': rc})
         self.sendStatus({'header': "elapsedTime=%0.6f\n" % self.elapsedTime})
         self._cancelTimers()
@@ -786,28 +786,29 @@ class RunProcess(object):
         if d:
             d.callback(rc)
         else:
-            log.msg("Hey, command %s finished twice" % self)
+            log.msg("Hey, command {0} finished twice".format(self))
 
     def failed(self, why):
         self._sendBuffers()
-        log.msg("RunProcess.failed: command failed: %s" % (why,))
+        log.msg("RunProcess.failed: command failed: {0}".format(why))
         self._cancelTimers()
         d = self.deferred
         self.deferred = None
         if d:
             d.errback(why)
         else:
-            log.msg("Hey, command %s finished twice" % self)
+            log.msg("Hey, command {0} finished twice".format(self))
 
     def doTimeout(self):
         self.ioTimeoutTimer = None
-        msg = "command timed out: %d seconds without output running %s" % (
-            self.timeout, self.fake_command)
+        msg = (
+            "command timed out: {0} seconds without output running {1}".format(
+            self.timeout, self.fake_command))
         self.kill(msg)
 
     def doMaxTimeout(self):
         self.maxTimeoutTimer = None
-        msg = "command timed out: %d seconds elapsed running %s" % (
+        msg = "command timed out: {0} seconds elapsed running {1}".format(
             self.maxTime, self.fake_command)
         self.kill(msg)
 
@@ -852,22 +853,22 @@ class RunProcess(object):
             sig = getattr(signal, "SIG" + interruptSignal, None)
 
             if sig is None:
-                log.msg("signal module is missing SIG%s" % interruptSignal)
+                log.msg("signal module is missing SIG{0}".format(interruptSignal))
             elif not hasattr(os, "kill"):
                 log.msg("os module is missing the 'kill' function")
             elif self.process.pgid is None:
                 log.msg("self.process has no pgid")
             else:
-                log.msg("trying to kill process group %d" %
-                        (self.process.pgid,))
+                log.msg("trying to kill process group {0}".format(
+                        self.process.pgid))
                 try:
                     os.killpg(self.process.pgid, sig)
-                    log.msg(" signal %s sent successfully" % sig)
+                    log.msg(" signal {0} sent successfully".format(sig))
                     self.process.pgid = None
                     hit = 1
                 except OSError:
-                    log.msg('failed to kill process group (ignored): %s' %
-                            (sys.exc_info()[1],))
+                    log.msg('failed to kill process group (ignored): {0}'.format(
+                            (sys.exc_info()[1])))
                     # probably no-such-process, maybe because there is no process
                     # group
                     pass
@@ -877,27 +878,27 @@ class RunProcess(object):
                 log.msg("interruptSignal==None, only pretending to kill child")
             elif self.process.pid is not None:
                 if interruptSignal == "TERM":
-                    log.msg("using TASKKILL PID /T to kill pid %s" %
-                            self.process.pid)
+                    log.msg("using TASKKILL PID /T to kill pid {0}".format(
+                            self.process.pid))
                     subprocess.check_call(
-                        "TASKKILL /PID %s /T" % self.process.pid)
-                    log.msg("taskkill'd pid %s" % self.process.pid)
+                        "TASKKILL /PID {0} /T".format(self.process.pid))
+                    log.msg("taskkill'd pid {0}".format(self.process.pid))
                     hit = 1
                 elif interruptSignal == "KILL":
-                    log.msg("using TASKKILL PID /F /T to kill pid %s" %
-                            self.process.pid)
+                    log.msg("using TASKKILL PID /F /T to kill pid {0}".format(
+                            self.process.pid))
                     subprocess.check_call(
-                        "TASKKILL /F /PID %s /T" % self.process.pid)
-                    log.msg("taskkill'd pid %s" % self.process.pid)
+                        "TASKKILL /F /PID {0} /T".format(self.process.pid))
+                    log.msg("taskkill'd pid {0}".format(self.process.pid))
                     hit = 1
 
         # try signalling the process itself (works on Windows too, sorta)
         if not hit:
             try:
-                log.msg("trying process.signalProcess('%s')" %
-                        (interruptSignal,))
+                log.msg("trying process.signalProcess('{0}')".format(
+                        interruptSignal))
                 self.process.signalProcess(interruptSignal)
-                log.msg(" signal %s sent successfully" % (interruptSignal,))
+                log.msg(" signal {0} sent successfully".format(interruptSignal))
                 hit = 1
             except OSError:
                 log.err("from process.signalProcess:")

--- a/worker/buildbot_worker/runprocess.py
+++ b/worker/buildbot_worker/runprocess.py
@@ -47,6 +47,7 @@ from twisted.python.win32 import quoteArguments
 
 from buildbot_worker import util
 from buildbot_worker.compat import bytes2unicode
+from buildbot_worker.compat import unicode2bytes
 from buildbot_worker.exceptions import AbandonChain
 
 if runtime.platformType == 'posix':
@@ -88,10 +89,10 @@ def shell_quote(cmd_list, unicode_encoding='utf-8'):
 
     def quote(e):
         if not e:
-            return '""'
+            return u'""'
         e = bytes2unicode(e, unicode_encoding)
         return pipes.quote(e)
-    return " ".join([quote(e) for e in cmd_list])
+    return u" ".join([quote(e) for e in cmd_list])
 
 
 class LogFileWatcher(object):
@@ -182,7 +183,7 @@ class RunProcessPP(protocol.ProcessProtocol):
 
     def __init__(self, command):
         self.command = command
-        self.pending_stdin = ""
+        self.pending_stdin = b""
         self.stdin_finished = False
         self.killed = False
 
@@ -465,6 +466,7 @@ class RunProcess(object):
         self.pp = RunProcessPP(self)
 
         self.using_comspec = False
+        self.command = unicode2bytes(self.command, encoding=self.builder.unicode_encoding)
         if isinstance(self.command, bytes):
             if runtime.platformType == 'win32':
                 # allow %COMSPEC% to have args
@@ -510,52 +512,53 @@ class RunProcess(object):
 
         # self.stdin is handled in RunProcessPP.connectionMade
 
-        log.msg(" " + display)
-        self._addToBuffers('header', display + "\n")
+        log.msg(u" " + display)
+        self._addToBuffers(u'header', display + u"\n")
 
         # then comes the secondary information
-        msg = " in dir {0}".format(self.workdir)
+        msg = u" in dir {0}".format(self.workdir)
         if self.timeout:
             if self.timeout == 1:
-                unit = "sec"
+                unit = u"sec"
             else:
-                unit = "secs"
-            msg += " (timeout {0} {1})".format(self.timeout, unit)
+                unit = u"secs"
+            msg += u" (timeout {0} {1})".format(self.timeout, unit)
         if self.maxTime:
             if self.maxTime == 1:
-                unit = "sec"
+                unit = u"sec"
             else:
-                unit = "secs"
-            msg += " (maxTime {0} {1})".format(self.maxTime, unit)
-        log.msg(" " + msg)
-        self._addToBuffers('header', msg + "\n")
+                unit = u"secs"
+            msg += u" (maxTime {0} {1})".format(self.maxTime, unit)
+        log.msg(u" " + msg)
+        self._addToBuffers(u'header', msg + u"\n")
 
         msg = " watching logfiles {0}".format(self.logfiles)
         log.msg(" " + msg)
-        self._addToBuffers('header', msg + "\n")
+        self._addToBuffers('header', msg + u"\n")
 
         # then the obfuscated command array for resolving unambiguity
-        msg = " argv: {0}".format(self.fake_command)
-        log.msg(" " + msg)
-        self._addToBuffers('header', msg + "\n")
+        msg = u" argv: {0}".format(self.fake_command)
+        log.msg(u" " + msg)
+        self._addToBuffers('header', msg + u"\n")
 
         # then the environment, since it sometimes causes problems
         if self.logEnviron:
-            msg = " environment:\n"
+            msg = u" environment:\n"
             env_names = sorted(self.environ.keys())
             for name in env_names:
-                msg += "  {0}={1}\n".format(name, self.environ[name])
-            log.msg(" environment:\n{0}".format(pprint.pformat(self.environ)))
-            self._addToBuffers('header', msg)
+                msg += u"  {0}={1}\n".format(bytes2unicode(name, encoding=self.builder.unicode_encoding),
+                                             bytes2unicode(self.environ[name], encoding=self.builder.unicode_encoding))
+            log.msg(u" environment:\n{0}".format(pprint.pformat(self.environ)))
+            self._addToBuffers(u'header', msg)
 
         if self.initialStdin:
-            msg = " writing {0} bytes to stdin".format(len(self.initialStdin))
-            log.msg(" " + msg)
-            self._addToBuffers('header', msg + "\n")
+            msg = u" writing {0} bytes to stdin".format(len(self.initialStdin))
+            log.msg(u" " + msg)
+            self._addToBuffers(u'header', msg + u"\n")
 
-        msg = " using PTY: {0}".format(bool(self.usePTY))
-        log.msg(" " + msg)
-        self._addToBuffers('header', msg + "\n")
+        msg = u" using PTY: {0}".format(bool(self.usePTY))
+        log.msg(u" " + msg)
+        self._addToBuffers(u'header', msg + u"\n")
 
         # put data into stdin and close it, if necessary.  This will be
         # buffered until connectionMade is called
@@ -613,7 +616,7 @@ class RunProcess(object):
         tf = NamedTemporaryFile(mode='w+', dir='.', suffix=".bat",
                                 delete=False)
         # echo off hides this cheat from the log files.
-        tf.write("@echo off\n")
+        tf.write(u"@echo off\n")
         if isinstance(self.command, (string_types, bytes)):
             tf.write(bytes2unicode(self.command, self.builder.unicode_encoding))
         else:

--- a/worker/buildbot_worker/runprocess.py
+++ b/worker/buildbot_worker/runprocess.py
@@ -649,7 +649,7 @@ class RunProcess(object):
         """
         retval = {}
         for logname in msg:
-            data = ""
+            data = u""
             for m in msg[logname]:
                 m = bytes2unicode(m, self.builder.unicode_encoding)
                 data += m

--- a/worker/buildbot_worker/runprocess.py
+++ b/worker/buildbot_worker/runprocess.py
@@ -46,7 +46,7 @@ from twisted.python import runtime
 from twisted.python.win32 import quoteArguments
 
 from buildbot_worker import util
-from buildbot_worker.compat import bytes2NativeString
+from buildbot_worker.compat import bytes2unicode
 from buildbot_worker.exceptions import AbandonChain
 
 if runtime.platformType == 'posix':
@@ -58,7 +58,7 @@ def win32_batch_quote(cmd_list, unicode_encoding='utf-8'):
     # Windows batch file. This is not quite the same as quoting it for the
     # shell, as cmd.exe doesn't support the %% escape in interactive mode.
     def escape_arg(arg):
-        arg = bytes2NativeString(arg, unicode_encoding)
+        arg = bytes2unicode(arg, unicode_encoding)
         arg = quoteArguments([arg])
         # escape shell special characters
         arg = re.sub(r'[@()^"<>&|]', r'^\g<0>', arg)
@@ -78,7 +78,7 @@ def shell_quote(cmd_list, unicode_encoding='utf-8'):
     # So:
     #  - use pipes.quote on UNIX, handling '' as a special case
     #  - use our own custom function on Windows
-    cmd_list = bytes2NativeString(cmd_list, unicode_encoding)
+    cmd_list = bytes2unicode(cmd_list, unicode_encoding)
 
     if runtime.platformType == 'win32':
         return win32_batch_quote(cmd_list, unicode_encoding)
@@ -88,7 +88,7 @@ def shell_quote(cmd_list, unicode_encoding='utf-8'):
     def quote(e):
         if not e:
             return '""'
-        e = bytes2NativeString(e, unicode_encoding)
+        e = bytes2unicode(e, unicode_encoding)
         return pipes.quote(e)
     return " ".join([quote(e) for e in cmd_list])
 
@@ -210,14 +210,14 @@ class RunProcessPP(protocol.ProcessProtocol):
     def outReceived(self, data):
         if self.debug:
             log.msg("RunProcessPP.outReceived")
-        data = bytes2NativeString(
+        data = bytes2unicode(
             data, self.command.builder.unicode_encoding)
         self.command.addStdout(data)
 
     def errReceived(self, data):
         if self.debug:
             log.msg("RunProcessPP.errReceived")
-        data = bytes2NativeString(
+        data = bytes2unicode(
             data, self.command.builder.unicode_encoding)
         self.command.addStderr(data)
 
@@ -484,7 +484,7 @@ class RunProcess(object):
             # Otherwise, we should run under COMSPEC (usually cmd.exe) to
             # handle path searching, etc.
             if (runtime.platformType == 'win32' and
-                not (bytes2NativeString(self.command[0]).lower().endswith(".exe") and
+                not (bytes2unicode(self.command[0]).lower().endswith(".exe") and
                      os.path.isabs(self.command[0]))):
                 # allow %COMSPEC% to have args
                 argv = os.environ['COMSPEC'].split()
@@ -498,7 +498,7 @@ class RunProcess(object):
             # isn't perfect
             display = shell_quote(self.fake_command, self.builder.unicode_encoding)
 
-        display = bytes2NativeString(display, self.builder.unicode_encoding)
+        display = bytes2unicode(display, self.builder.unicode_encoding)
 
         # $PWD usually indicates the current directory; spawnProcess may not
         # update this value, though, so we set it explicitly here.  This causes
@@ -613,7 +613,7 @@ class RunProcess(object):
         # echo off hides this cheat from the log files.
         tf.write("@echo off\n")
         if isinstance(self.command, (string_types, bytes)):
-            tf.write(bytes2NativeString(self.command))
+            tf.write(bytes2unicode(self.command))
         else:
             tf.write(win32_batch_quote(self.command))
         tf.close()
@@ -649,7 +649,7 @@ class RunProcess(object):
         for logname in msg:
             data = ""
             for m in msg[logname]:
-                m = bytes2NativeString(m, self.builder.unicode_encoding)
+                m = bytes2unicode(m, self.builder.unicode_encoding)
                 data += m
             if isinstance(logname, tuple) and logname[0] == 'log':
                 retval['log'] = (logname[1], data)

--- a/worker/buildbot_worker/runprocess.py
+++ b/worker/buildbot_worker/runprocess.py
@@ -78,7 +78,8 @@ def shell_quote(cmd_list, unicode_encoding='utf-8'):
     # So:
     #  - use pipes.quote on UNIX, handling '' as a special case
     #  - use our own custom function on Windows
-    cmd_list = bytes2unicode(cmd_list, unicode_encoding)
+    if isinstance(cmd_list, bytes):
+        cmd_list = bytes2unicode(cmd_list, unicode_encoding)
 
     if runtime.platformType == 'win32':
         return win32_batch_quote(cmd_list, unicode_encoding)

--- a/worker/buildbot_worker/runprocess.py
+++ b/worker/buildbot_worker/runprocess.py
@@ -485,7 +485,8 @@ class RunProcess(object):
             # Otherwise, we should run under COMSPEC (usually cmd.exe) to
             # handle path searching, etc.
             if (runtime.platformType == 'win32' and
-                not (bytes2unicode(self.command[0]).lower().endswith(".exe") and
+                not (bytes2unicode(self.command[0],
+                     self.builder.unicode_encoding).lower().endswith(".exe") and
                      os.path.isabs(self.command[0]))):
                 # allow %COMSPEC% to have args
                 argv = os.environ['COMSPEC'].split()
@@ -614,7 +615,7 @@ class RunProcess(object):
         # echo off hides this cheat from the log files.
         tf.write("@echo off\n")
         if isinstance(self.command, (string_types, bytes)):
-            tf.write(bytes2unicode(self.command))
+            tf.write(bytes2unicode(self.command, self.builder.unicode_encoding))
         else:
             tf.write(win32_batch_quote(self.command))
         tf.close()

--- a/worker/buildbot_worker/test/unit/test_bot.py
+++ b/worker/buildbot_worker/test/unit/test_bot.py
@@ -33,6 +33,7 @@ from twisted.trial import unittest
 import buildbot_worker
 from buildbot_worker import base
 from buildbot_worker import pb
+from buildbot_worker.commands.base import Command
 from buildbot_worker.test.fake.remote import FakeRemote
 from buildbot_worker.test.fake.runprocess import Expect
 from buildbot_worker.test.util import command
@@ -160,6 +161,8 @@ class TestBot(unittest.TestCase):
                 self.assertEqual(
                     id(workerforbuilders['my']), id(builders['mybld']))
                 workerforbuilders['your'] = builders['yourbld']
+                self.assertTrue(repr(workerforbuilders['your']).startswith(
+                                 "<WorkerForBuilder 'yourbld' at "))
             d.addCallback(check)
             return d
         d.addCallback(add_your)
@@ -264,6 +267,10 @@ class TestWorkerForBuilder(command.CommandTestMixin, unittest.TestCase):
 
     def test_print(self):
         return self.wfb.callRemote("print", "Hello, WorkerForBuilder.")
+
+    def test_printWithCommand(self):
+        self.wfb.original.command = Command("builder", "1", ["arg1", "arg2"])
+        return self.wfb.callRemote("print", "Hello again, WorkerForBuilder.")
 
     def test_setMaster(self):
         # not much to check here - what the WorkerForBuilder does with the
@@ -395,6 +402,18 @@ class TestWorkerForBuilder(command.CommandTestMixin, unittest.TestCase):
                                        "13", "shell", dict())
 
         yield self.assertFailure(do_start(), ValueError)
+
+    @defer.inlineCallbacks
+    def test_startCommand_invalid_command(self):
+        # set up a fake step to receive updates
+        st = FakeStep()
+
+        def do_start():
+            return self.wfb.callRemote("startCommand", FakeRemote(st),
+                                       "13", "invalid command", dict())
+
+        unknownCommand = yield self.assertFailure(do_start(), base.UnknownCommand)
+        self.assertEqual(str(unknownCommand), "unrecognized WorkerCommand 'invalid command'")
 
 
 class TestBotFactory(unittest.TestCase):

--- a/worker/buildbot_worker/test/unit/test_runprocess.py
+++ b/worker/buildbot_worker/test/unit/test_runprocess.py
@@ -462,6 +462,25 @@ class TestRunProcess(BasedirMixin, unittest.TestCase):
                           runprocess.RunProcess(b, stdoutCommand('hello'), self.basedir,
                                                 environ={"BUILD_NUMBER": 13}))
 
+    def test_spawnAsBatch(self):
+
+        def spawnProcess(processProtocol, executable, args=(), env=None,
+                      path=None, uid=None, gid=None, usePTY=False, childFDs=None):
+            self.assertTrue(args[0].lower().endswith("cmd.exe"),
+                            "{0} is not cmd.exe".format(args[0]))
+
+        self.patch(runprocess.reactor, "spawnProcess", spawnProcess)
+        tempEnviron = os.environ.copy()
+        if 'COMSPEC' not in tempEnviron:
+            tempEnviron['COMSPEC'] = "cmd.exe"
+        self.patch(os, "environ", tempEnviron)
+        b = FakeWorkerForBuilder(self.basedir)
+        s = runprocess.RunProcess(b, "dir c:/", self.basedir)
+        s.pp = runprocess.RunProcessPP(s)
+        s.deferred = defer.Deferred()
+        d = s._spawnAsBatch(s.pp, s.command, "args", tempEnviron, "path", False)
+        return d
+
 
 class TestPOSIXKilling(BasedirMixin, unittest.TestCase):
 

--- a/worker/buildbot_worker/test/unit/test_runprocess.py
+++ b/worker/buildbot_worker/test/unit/test_runprocess.py
@@ -462,7 +462,7 @@ class TestRunProcess(BasedirMixin, unittest.TestCase):
                           runprocess.RunProcess(b, stdoutCommand('hello'), self.basedir,
                                                 environ={"BUILD_NUMBER": 13}))
 
-    def test_spawnAsBatch(self):
+    def _test_spawnAsBatch(self, cmd, comspec):
 
         def spawnProcess(processProtocol, executable, args=(), env=None,
                       path=None, uid=None, gid=None, usePTY=False, childFDs=None):
@@ -472,14 +472,20 @@ class TestRunProcess(BasedirMixin, unittest.TestCase):
         self.patch(runprocess.reactor, "spawnProcess", spawnProcess)
         tempEnviron = os.environ.copy()
         if 'COMSPEC' not in tempEnviron:
-            tempEnviron['COMSPEC'] = "cmd.exe"
+            tempEnviron['COMSPEC'] = comspec
         self.patch(os, "environ", tempEnviron)
         b = FakeWorkerForBuilder(self.basedir)
-        s = runprocess.RunProcess(b, "dir c:/", self.basedir)
+        s = runprocess.RunProcess(b, cmd, self.basedir)
         s.pp = runprocess.RunProcessPP(s)
         s.deferred = defer.Deferred()
         d = s._spawnAsBatch(s.pp, s.command, "args", tempEnviron, "path", False)
         return d
+
+    def test_spawnAsBatchCommandString(self):
+        return self._test_spawnAsBatch("dir c:/", "cmd.exe")
+
+    def test_spawnAsBatchCommandList(self):
+        return self._test_spawnAsBatch(stdoutCommand('hello'), "cmd.exe /c")
 
 
 class TestPOSIXKilling(BasedirMixin, unittest.TestCase):
@@ -810,6 +816,15 @@ class TestLogging(BasedirMixin, unittest.TestCase):
         data = "x" * (runprocess.RunProcess.BUFFER_SIZE + 1)
         s._addToBuffers('stdout', data)
         self.assertEqual(len(b.updates), 1)
+
+    def testSendLog(self):
+        b = FakeWorkerForBuilder(self.basedir)
+        s = runprocess.RunProcess(b, stdoutCommand('hello'), self.basedir)
+        s._addToBuffers(('log', 'stdout'), 'hello ')
+        s._sendBuffers()
+        self.assertEqual(b.updates, [
+            {'log': ('stdout', 'hello ')},
+        ])
 
 
 class TestLogFileWatcher(BasedirMixin, unittest.TestCase):

